Try resetting the GPU process if rendering commands time out

kkinnunen-apple opened the pull request:
#### 3dcfec1cff82ddf5d5c0cd6351c4366a50a139b5
<pre>
Try resetting the GPU process if rendering commands time out
<a href="https://bugs.webkit.org/show_bug.cgi?id=274934">https://bugs.webkit.org/show_bug.cgi?id=274934</a>
<a href="https://rdar.apple.com/125143181">rdar://125143181</a>

Reviewed by Simon Fraser and Cameron McCormack.

Terminate GPU process if remote rendering message sends time out.
If the GPU procsess side threads are not processing the messages,
the sends start failing once the IPC stream buffer fills up.
Mark the RemoteRenderingBackendProxy as non-responsive upon first
timeout.
Send unresponsiveness notification to the UI process.
Terminate the GPU process upon receiving the notification.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::waitForMessage):
* Source/WebKit/Scripts/webkit/messages.py:
(handler_function):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::drawInContext):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::createContextAndPaintContents):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::childConnectionDidBecomeUnresponsive):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::gpuProcessConnectionDidBecomeUnresponsive):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::didBecomeUnresponsive):
(WebKit::GPUProcessConnection::didClose):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::send):
(WebKit::RemoteDisplayListRecorderProxy::connection const):
(WebKit::RemoteDisplayListRecorderProxy::didBecomeUnresponsive const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::send):
(WebKit::RemoteImageBufferProxy::sendSync):
(WebKit::RemoteImageBufferProxy::connection const):
(WebKit::RemoteImageBufferProxy::didBecomeUnresponsive const):
(WebKit::RemoteImageBufferProxy::ensureBackend const):
(WebKit::RemoteImageBufferProxy::streamConnection const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.cpp:
(WebKit::RemoteImageBufferSetProxy::send):
(WebKit::RemoteImageBufferSetProxy::sendSync):
(WebKit::RemoteImageBufferSetProxy::connection const):
(WebKit::RemoteImageBufferSetProxy::didBecomeUnresponsive const):
(WebKit::RemoteImageBufferSetProxy::flushFrontBufferAsync):
(WebKit::RemoteImageBufferSetProxy::willPrepareForDisplay):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::~RemoteRenderingBackendProxy):
(WebKit::RemoteRenderingBackendProxy::ensureGPUProcessConnection):
(WebKit::RemoteRenderingBackendProxy::send):
(WebKit::RemoteRenderingBackendProxy::sendSync):
(WebKit::RemoteRenderingBackendProxy::didClose):
(WebKit::RemoteRenderingBackendProxy::didBecomeUnresponsive):
(WebKit::RemoteRenderingBackendProxy::disconnectGPUProcess):
(WebKit::RemoteRenderingBackendProxy::releaseImageBuffer):
(WebKit::RemoteRenderingBackendProxy::releaseRemoteImageBufferSet):
(WebKit::RemoteRenderingBackendProxy::releaseAllDrawingResources):
(WebKit::RemoteRenderingBackendProxy::releaseRenderingResource):
(WebKit::RemoteRenderingBackendProxy::releaseAllImageResources):
(WebKit::RemoteRenderingBackendProxy::prepareImageBufferSetsForDisplay):
(WebKit::RemoteRenderingBackendProxy::markSurfacesVolatile):
(WebKit::RemoteRenderingBackendProxy::finalizeRenderingUpdate):
(WebKit::RemoteRenderingBackendProxy::didPaintLayers):
(WebKit::RemoteRenderingBackendProxy::connection):
(WebKit::RemoteRenderingBackendProxy::didInitialize):
(WebKit::Function&lt;bool):
(WebKit::RemoteRenderingBackendProxy::streamConnection): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
(WebKit::RemoteRenderingBackendProxy::isGPUProcessConnectionClosed const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createBarcodeDetector const):
(WebKit::WebChromeClient::getBarcodeDetectorSupportedFormats const):
(WebKit::WebChromeClient::createFaceDetector const):
(WebKit::WebChromeClient::createTextDetector const):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
(WebKit::RemoteLayerTreeDrawingArea::didCompleteRenderingUpdateDisplayFlush):
(WebKit::RemoteLayerTreeDrawingArea::BackingStoreFlusher::flush):
(WebKit::RemoteLayerTreeDrawingArea::didCompleteRenderingUpdateDisplay): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didFlushLayerTreeAtTime):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::gpuProcessConnectionClosed):
(WebKit::WebProcess::gpuProcessConnectionDidBecomeUnresponsive):
* Source/WebKit/WebProcess/WebProcess.h:

Canonical link: <a href="https://commits.webkit.org/279833@main">https://commits.webkit.org/279833@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7fab4dae7fa9f3ef5c0375181e8d720309fe4d16

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54671 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34093 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7247 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57951 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5404 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41648 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5413 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44285 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3648 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56766 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32237 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47353 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25410 "Found 1 new API test failure: /WPE/TestWebProcessExtensions:/webkit/WebKitWebView/web-process-crashed (failure)") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/54499 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29033 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3544 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/50769 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4901 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59541 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/29912 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5049 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51707 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31057 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47439 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51093 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/11980 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32046 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8089 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/30842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
```diff
@@ -549,9 +549,11 @@ void GPUProcessProxy::processIsReadyToExit()
+void GPUProcessProxy::requestTermination(ProcessTerminationReason reason)
 {
-    processIsReadyToExit();
+    RELEASE_LOG(Process, "%p - GPUProcessProxy::requestTermination:", this);
```
Should we include the `reason` in the log message? That might be helpful for analyzing GPU process terminations later.

heycam commented on Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
```diff
@@ -549,9 +549,11 @@ void GPUProcessProxy::processIsReadyToExit()
-void GPUProcessProxy::terminateForTesting()
+void GPUProcessProxy::requestTermination(ProcessTerminationReason reason)
 {
-    processIsReadyToExit();
```
There are no other callers of `processIsReadyToExit`. Can we remove it?

kkinnunen-apple replied:
processIsReadyToExit is a method send from GPUP to UI

heycam commented on Source/WebKit/UIProcess/WebProcessProxy.h:
```diff
@@ -776,6 +777,7 @@ class WebProcessProxy : public AuxiliaryProcessProxy {
     mutable String m_environmentIdentifier;
 #if ENABLE(GPU_PROCESS)
     mutable std::optional<GPUProcessPreferencesForWebProcess> m_preferencesForGPUProcess;
+    bool m_hasTerminatedGPUProcess { false };
```
We do not use this variable. Remove it?

heycam commented on Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
```diff
@@ -70,22 +71,35 @@ RemoteDisplayListRecorderProxy::RemoteDisplayListRecorderProxy(RemoteRenderingBa
-            renderingBackendIdentifier.toUInt64(), IPC::description(T::name()).characters(), IPC::errorAsString(result).characters());
+        RELEASE_LOG(RemoteLayerBuffers, "RemoteDisplayListRecorderProxy::send - failed, name:%" PUBLIC_LOG_STRING ", error:%" PUBLIC_LOG_STRING,
+            IPC::description(T::name()).characters(), IPC::errorAsString(result).characters());
+        didBecomeUnresponsive();
```
Should we only call this for `Error::Timeout`? I guess they're all errors that mean we can't really trust the state of the GPU process and we should probably kill it eagerly anyway (instead of then waiting for some subsequent timeout). But the name `didBecomeUnresponsive` doesn't match the other error types.

(And similar for the other rendering-related connection client classes we're adding this handling to.)

kkinnunen-apple replied:
I think it's appropriate.. Any error makes the connection unresponsive.

heycam commented on Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
```diff
@@ -81,35 +81,45 @@ void RemoteImageBufferProxy::assertDispatcherIsCurrent() const
+
+void RemoteImageBufferProxy::didBecomeUnresponsive() const
+{
+    m_remoteRenderingBackendProxy->didBecomeUnresponsive();
```
Do we need to null check the weak `m_remoteRenderingBackendProxy`?

heycam commented on Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.cpp:
```diff
@@ -111,36 +111,47 @@ class RemoteImageBufferSetProxyFlusher final : public ThreadSafeImageBufferSetFl
+
+void RemoteImageBufferSetProxy::didBecomeUnresponsive() const
+{
+    m_remoteRenderingBackendProxy->didBecomeUnresponsive();
```
And here.

heycam commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
```diff
@@ -121,27 +121,34 @@ void RemoteRenderingBackendProxy::ensureGPUProcessConnection()
-    auto result = streamConnection().send(std::forward<T>(message), destination, defaultTimeout);
+    RefPtr connection = this->connection();
+    if (UNLIKELY(!connection))
+        return IPC::Error::Timeout;
```
Is `Timeout` the right error here? Would `InvalidConnection` be more accurate (like you use in `sendSync`)?

heycam commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
```diff
@@ -153,15 +160,30 @@ void RemoteRenderingBackendProxy::didClose(IPC::Connection&)
+{
+    if (!m_isResponsive)
+        return;
+    RELEASE_LOG(RemoteLayerBuffers, "[renderingBackend=%" PRIu64 "] RemoteRenderingBackendProxy::didBecomeUnresponsive",  m_identifier.toUInt64());
```
Nit: two spaces before `m_identifier`.

(And not for you to fix, but this file should probably get its own log channel instead of using `RemoteLayerBuffers`.)

heycam commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -1026,7 +1026,11 @@ RefPtr<WebCore::ShapeDetection::BarcodeDetector> WebChromeClient::createBarcodeD
     auto page = protectedPage();
     auto& remoteRenderingBackendProxy = page->ensureRemoteRenderingBackendProxy();
-    return ShapeDetection::RemoteBarcodeDetectorProxy::create(remoteRenderingBackendProxy.streamConnection(), remoteRenderingBackendProxy.renderingBackendIdentifier(), ShapeDetectionIdentifier::generate(), barcodeDetectorOptions);
+    // FIXME: Does not work when the GPUP crashes.
```
Can you file a followup bug for this work?

heycam reviewed (approved):
r=me with comments addressed.

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
```diff
@@ -80,9 +80,11 @@
+    bool flushAndCollectHandles(HashMap<RemoteImageBufferSetIdentifier, std::unique_ptr<BufferSetBackendHandle>>&) final
     {
-        m_fence->waitFor(delegatedContentsFinishedTimeout);
+        if (m_fence->waitFor(delegatedContentsFinishedTimeout))
```
`return m_fence->waitFor()` ?

smfr commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1083,6 +1083,13 @@ void WebProcessProxy::createGPUProcessConnection(IPC::Connection::Handle&& conne
 
+void WebProcessProxy::gpuProcessConnectionDidBecomeUnresponsive()
+{
+    WEBPROCESSPROXY_RELEASE_LOG_ERROR(Process, "gpuProcessConnectionDidBecomeUnresponsive");
```
Does this log the PID of the unresponsive GPU process? It should.

kkinnunen-apple replied:
Later calls in the callstack will report that (GPUProcessProxy::didBecomeUnresponsive)

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
```diff
@@ -489,21 +489,25 @@
+    for (auto& flusher : flushers) {
+        flushSucceeded = flusher->flushAndCollectHandles(handles);
+        if (!flushSucceeded)
+            break;
```
Should we release log here?

kkinnunen-apple committed:
Try resetting the GPU process if rendering commands time out
https://bugs.webkit.org/show_bug.cgi?id=274934
rdar://125143181

Reviewed by Simon Fraser and Cameron McCormack.

Terminate GPU process if remote rendering message sends time out.
If the GPU procsess side threads are not processing the messages,
the sends start failing once the IPC stream buffer fills up.
Mark the RemoteRenderingBackendProxy as non-responsive upon first
timeout.
Send unresponsiveness notification to the UI process.
Terminate the GPU process upon receiving the notification.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::waitForMessage):
* Source/WebKit/Scripts/webkit/messages.py:
(handler_function):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::drawInContext):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::createContextAndPaintContents):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::childConnectionDidBecomeUnresponsive):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::gpuProcessConnectionDidBecomeUnresponsive):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::didBecomeUnresponsive):
(WebKit::GPUProcessConnection::didClose):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::send):
(WebKit::RemoteDisplayListRecorderProxy::connection const):
(WebKit::RemoteDisplayListRecorderProxy::didBecomeUnresponsive const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::send):
(WebKit::RemoteImageBufferProxy::sendSync):
(WebKit::RemoteImageBufferProxy::connection const):
(WebKit::RemoteImageBufferProxy::didBecomeUnresponsive const):
(WebKit::RemoteImageBufferProxy::ensureBackend const):
(WebKit::RemoteImageBufferProxy::streamConnection const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.cpp:
(WebKit::RemoteImageBufferSetProxy::send):
(WebKit::RemoteImageBufferSetProxy::sendSync):
(WebKit::RemoteImageBufferSetProxy::connection const):
(WebKit::RemoteImageBufferSetProxy::didBecomeUnresponsive const):
(WebKit::RemoteImageBufferSetProxy::flushFrontBufferAsync):
(WebKit::RemoteImageBufferSetProxy::willPrepareForDisplay):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::~RemoteRenderingBackendProxy):
(WebKit::RemoteRenderingBackendProxy::ensureGPUProcessConnection):
(WebKit::RemoteRenderingBackendProxy::send):
(WebKit::RemoteRenderingBackendProxy::sendSync):
(WebKit::RemoteRenderingBackendProxy::didClose):
(WebKit::RemoteRenderingBackendProxy::didBecomeUnresponsive):
(WebKit::RemoteRenderingBackendProxy::disconnectGPUProcess):
(WebKit::RemoteRenderingBackendProxy::releaseImageBuffer):
(WebKit::RemoteRenderingBackendProxy::releaseRemoteImageBufferSet):
(WebKit::RemoteRenderingBackendProxy::releaseAllDrawingResources):
(WebKit::RemoteRenderingBackendProxy::releaseRenderingResource):
(WebKit::RemoteRenderingBackendProxy::releaseAllImageResources):
(WebKit::RemoteRenderingBackendProxy::prepareImageBufferSetsForDisplay):
(WebKit::RemoteRenderingBackendProxy::markSurfacesVolatile):
(WebKit::RemoteRenderingBackendProxy::finalizeRenderingUpdate):
(WebKit::RemoteRenderingBackendProxy::didPaintLayers):
(WebKit::RemoteRenderingBackendProxy::connection):
(WebKit::RemoteRenderingBackendProxy::didInitialize):
(WebKit::Function<bool):
(WebKit::RemoteRenderingBackendProxy::streamConnection): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
(WebKit::RemoteRenderingBackendProxy::isGPUProcessConnectionClosed const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createBarcodeDetector const):
(WebKit::WebChromeClient::getBarcodeDetectorSupportedFormats const):
(WebKit::WebChromeClient::createFaceDetector const):
(WebKit::WebChromeClient::createTextDetector const):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
(WebKit::RemoteLayerTreeDrawingArea::didCompleteRenderingUpdateDisplayFlush):
(WebKit::RemoteLayerTreeDrawingArea::BackingStoreFlusher::flush):
(WebKit::RemoteLayerTreeDrawingArea::didCompleteRenderingUpdateDisplay): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didFlushLayerTreeAtTime):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::gpuProcessConnectionClosed):
(WebKit::WebProcess::gpuProcessConnectionDidBecomeUnresponsive):
* Source/WebKit/WebProcess/WebProcess.h:

Canonical link: https://commits.webkit.org/279833@main
